usr/src/uts/common/fs/zfs/zap.c

```diff
@@ -505,6 +505,16 @@ zap_get_leaf_byblk(zap_t *zap, uint64_t blkid, dmu_tx_t *tx, krw_t lt,
 
 	ASSERT(RW_LOCK_HELD(&zap->zap_rwlock));
 
+	/*
+	 * If system crashed just after dmu_free_long_range in zfs_rmnode, we
+	 * would be left with an empty xattr dir in delete queue. blkid=0
+	 * would be passed in when doing zfs_purgedir. If that's the case we
+	 * should just return immediately. The underlying objects should
+	 * already be freed, so this should be perfectly fine.
+	 */
+	if (blkid == 0)
```

Comment:
There was some discussion at the time about how to detect this issue earlier since we didn't want modify `zap_get_leaf_byblk()`.  However, we never settled on a better solution and historically we've had quite a few people encounter this issue so we needed a fix.  I'm all for a better solution.